examples/TestAppUwp/NodeDssSignaler.cs

```diff
@@ -289,110 +291,119 @@ public Task SendMessageAsync(Message message)
-                _cancellationTokenSource = null;
+                lock (_pollingLock)
+                {
+                    _isPolling = false;
+                    _cancellationTokenSource.Dispose();
+                    _cancellationTokenSource = null;
+                }
+                RaisePropertyChanged("IsPolling");
+                Interlocked.Exchange(ref _connectedEventFired, 0);
+                OnDisconnect?.Invoke();
                 OnPollingDone?.Invoke();
+                masterTokenSource.Dispose();
             });
 
             // Prepare the repeating poll task.
             // In order to poll at the specified frequency but also avoid overlapping requests,
             // use a repeating task which re-schedule itself on completion, either immediately
             // if the polling delay is elapsed, or at a later time otherwise.
-            Action nextAction = null;
-            nextAction = () =>
+            async void PollServerOnce()
             {
-                token.ThrowIfCancellationRequested();
-
-                // Send GET request to DSS server.
-                // In order to avoid exceptions in GetStreamAsync() when the server returns a non-success status code (e.g. 404),
-                // first get the HTTP headers, check the status code, then if successful wait for content.
-                lastPollTimeTicks = DateTime.UtcNow.Ticks;
-                _httpClient.GetAsync(requestUri, HttpCompletionOption.ResponseHeadersRead).ContinueWith((getTask) =>
+                try
                 {
-                    if (getTask.Exception != null)
+                    masterToken.ThrowIfCancellationRequested();
+
+                    // Send GET request to DSS server.
+                    lastPollTimeTicks = DateTime.UtcNow.Ticks;
+                    HttpResponseMessage response = await _httpClient.GetAsync(requestUri, HttpCompletionOption.ResponseHeadersRead);
+
+                    // On first successful HTTP request, raise the connected event
+                    if (0 == Interlocked.Exchange(ref _connectedEventFired, 1))
                     {
-                        OnFailure?.Invoke(getTask.Exception);
-                        return;
+                        OnConnect?.Invoke();
                     }
 
-                    token.ThrowIfCancellationRequested();
+                    masterToken.ThrowIfCancellationRequested();
 
-                    HttpResponseMessage response = getTask.Result;
+                    // In order to avoid exceptions in GetStreamAsync() when the server returns a non-success status code (e.g. 404),
+                    // first get the HTTP headers, check the status code, then if successful wait for content.
                     if (response.IsSuccessStatusCode)
                     {
-                        response.Content.ReadAsStringAsync().ContinueWith((readTask) =>
+                        string jsonMsg = await response.Content.ReadAsStringAsync();
```

Comment:
nit: should abort/discard this task if `masterToken` is canceled - though I would expect getting a string here to be reasonably fast (dunno what happens under the hood), so maybe it's fine as it is.

Reply:
Not sure I understand. `ReadAsStringAsync()` doesn't take a cancellation token, and the `masterToken` is tested right before and right after the call. How do you expect to do better than that?

Reply:
There are a number of ways to work around lack of a CancellationToken param, see https://www.planetgeek.ch/2016/12/02/async-method-without-cancellation-support-oh-my/ for example. But might not be strictly needed here, there should be no dire consequences if the task stays alive a few ms longer.

Reply:
Right, good resource, I didn't know about that pattern. However I'd close as won't fix here though because the `node-dss` messages are tiny (few KB at most for offer/answer), and `ReadAsStringAsync()` most likely immediately returns (note that we already read the HTTP header above, so we're reading only the content here, and by the time we reached that code there's an even higher chance the body of the HTTP message was received). In the linked example, `Task.WhenAny()` will indeed return and present a completed task to the outer code waiting for it, but the actual task doing the work internally (not the one with delay) doesn't seem to be cancelled, so is just running in the background until it's completed and discarded. So the only thing that changes is how fast is the outermost `Task`; turns out `StartPollingAsync()` does not return that `Task` anyway so nobody can wait for it, so this makes no difference here.

Reply:
> but the actual task doing the work internally (not the one with delay) doesn't seem to be cancelled

FWIW I guess that you could kill that task too by disposing the HttpContent once you get control back. Anyway, let's move on!